### ISSUE TITLE
Feature Request: Global Statusline

### ISSUE THREAD

**Issue by marcotrosi** [vim/vim#10153]
NeoVim recently introduced the global statusline, which sounds great to me, because it would allow me to display global information like the current working directory, global settings, git/project related information in the global statusline and buffer/file related information in the window statusline.

It also should be possible to turn on/off the global and the window statuslines individually.

Remark: I'm not a NeoVim user but I know that the multiline statusline is a thing that many Vimmers are wishing for. And now with the concept of a global statusline it is almost like multiline because we would have 2 statuslines. And with a single window it would even look like a 2 line statusline.

I checked if such requests exist, but I couldn't find any. Feel free to link.

EDIT: I used NeoVim only as a trigger, I want what I described in the images below, and not the exact NeoVim solution.

**Comment by marcotrosi** [comment]
Here the link to the proposal on nvim side
https://github.com/neovim/neovim/issues/9342

**Comment by brammool** [comment]
So it is no longer clear what the contents of window is?  I would think a global statusline would be an extra line above the command line, not replacing the existing statuslines but moving some items there.

**Comment by marcotrosi** [comment]
> So it is no longer clear what the contents of window is? I would think a global statusline would be an extra line above the command line, not replacing the existing statuslines but moving some items there.

<img width="853" alt="global statusline" src="https://user-images.githubusercontent.com/8596795/162824824-074fe651-9524-4577-8745-e641763c6a51.png">

The question I don't understand, but yes the global statusline is an extra line (or more) above the commandline.
I think it would be nice if one could configure the global statusline independently from the window statuslines.
And also deciding if the global statusline is visible or not, and if window statuslines are visible or not, or only when the window is focused.

an example configuration could be...
the global statusline shows current working directory and git branch name
the window statuslines show the filename number of the current line, total number of lines

but of course one can configure whatever is desired.
I could also configure to only have the global statusline, and only show the information of the currently focused window.

**Comment by marcotrosi** [comment]
Here 2 more images as described in my previous message.

<img width="846" alt="focused window" src="https://user-images.githubusercontent.com/8596795/162828343-4c7ec519-f4ab-4032-87ee-025f62c8c77d.png">

<img width="850" alt="only global" src="https://user-images.githubusercontent.com/8596795/162828354-fda87af5-d7b4-4c34-b024-a640b409e9d6.png">

**Comment by marcotrosi** [comment]
I noticed that when someone has 4 splits or more, that then the statusline is taking either too much space vertically, or the statusline is too short and not all information fits. And a global statusline would solve that issue, because it is 1 line only and has the full width. But of course it all depends on the configuration as the images show.

**Comment by marcotrosi** [comment]
> So it is no longer clear what the contents of window is? I would think a global statusline would be an extra line above the command line, not replacing the existing statuslines but moving some items there.

Ah you mean that in vertical splits we wouldn't be able to see where a window ends and the next split window starts?
Okay now I see that my vertical space argument is not valid, because in Terminal we still need a full line even if we only want to display a thin separator line as shown in the image. But still visually clean and gives the impression of more space. 
 
 So this would mean that in the images above where a window has no statusline, the space where the statusline would have been gets replaced with a separator line? Maybe with box drawing characters?
 
 Yes that sounds right and logical to me, thanks @brammool for pointing it out.

**Comment by monkoose** [comment]
@marcotrosi It's not how global statusline currently works in neovim. It's just one statusline and other windows just have borders not another statuslines.

**Comment by monkoose** [comment]
> So it is no longer clear what the contents of window is? I would think a global statusline would be an extra line above the command line, not replacing the existing statuslines but moving some items there.

And why only above command line? Maybe as first line of the screen would be better or an option where to place it?

**Comment by marcotrosi** [comment]
> @marcotrosi It's not how global statusline currently works in neovim. It's just one statusline and other windows just have borders not another statuslines. So Bram question was actually on point.

that is correct, but I want it differently, I didn't want the NeoVim solution. And with my general approach users can configure whatever they need. If they want window statuslines they can have them, if they dont they dont.

**Comment by marcotrosi** [comment]
> > So it is no longer clear what the contents of window is? I would think a global statusline would be an extra line above the command line, not replacing the existing statuslines but moving some items there.
> 
> And why only above command line? Maybe as first line of the screen would be better or an option where to place it?

This is theoretically possible, it would be then above the tabline. But to be honest, this is already a quite heavy feature request, and I don't want to bother Bram too much. So I would like to keep it simple and increase the chances of getting such feature.

**Comment by skysky97** [comment]
The nvim solution is better to me. And I even want  putting global status line and cmd line together into a single line, just like vscode.

**Comment by ubaldot** [comment]
I was wondering the same [recently](https://github.com/vim/vim/discussions/15702#discussioncomment-10709040). 

If I have to express my personal preference, I would go for this:

https://github.com/vim/vim/discussions/15702#discussioncomment-10709040

Tidy and clean. That also remind me the macos menu bar that switches depending on who is active at that point in time. 

Note that if you have a crowded statusline, then the window content may not be clear anyway. In that case, is better to have a global statusline so at least you know what is one window content (i.e. the active window content). Plus, and I may be wrong, it looks like common practice among IDE:s to have a unique statusline that switch its content depending on the active window, I think we could easily align.

Finally, note that the request does not imply any changes of existing behaviour, so that for those not interested nothing will change. All in all I don't think it is an unreasonable request, given also the number of thumbs up on top of the first post that you seldom see. :) 

My two cents.

**Comment by mazunki** [comment]
I've also wanted to separate buffer-specific information from information which concern the state of vim itself. My two cents is that I'd want to move the global statusline to the very top. I seldom use tabs nor splits, and still want this.

**Comment by marcotrosi** [comment]
Hello @chrisbra,

I noticed you put recently some effort into the statusline area of the code. Maybe we can talk again about the global statusline feature?

a single global statusline that can be configured and turned on/off. Of course while keeping the normal per window statusline feature that also can be configured and turned on/off as of today.
See various images above.

this way we can display global information that is valid for all windows or the session in the global statusline. And we can display information in the window statusline that is related to the buffer or window displayed in that window.

and maybe an easy way to automatically turn on/off the per window statusline when the cursor is in that window (focused window)

Best Regards,
--marco

**Comment by h-east** [comment]
It's been possible for quite some time to display information such as the current directory at the top by properly configuring 'tabline', but wouldn't that be sufficient?

In the sample below, the tab page information is kept to a minimum, and the current directory and Git status are displayed in the remaining space.

tabline_example.vim:

```vim
vim9script

g:git_ps1_str = ''
set showtabline=2
if has("gui_running")
  set guioptions-=e
endif
&tabline = "%!MyTabLine()"
def g:MyTabLine(): string
  var tabpages_info = ''
  if tabpagenr('$') > 1
    tabpages_info = $"%#TabLine#[{tabpagenr()}/{tabpagenr('$')}]%#TabLineFill# "
  endif
  return $"{tabpages_info}%<cwd: {fnamemodify(getcwd(), ":~")} {g:git_ps1_str}"
enddef

def GitStatus(): string
  silent var branch = system('git branch --show-current 2>/dev/null')
  if v:shell_error != 0
    return ''
  endif
  branch = substitute(branch, '\n', '', 'g')
  silent var dirty = system('git status --porcelain -uno 2>/dev/null') !=# '' ? ' *' : ''
  return $'({branch}{dirty})'
enddef

def UpdateGitPS1()
  g:git_ps1_str = GitStatus()
enddef

augroup GitPS1
  autocmd!
  autocmd VimEnter,BufEnter,BufWritePost,FocusGained,DirChanged * UpdateGitPS1()
augroup END

# vim: set et ts=2 sw=2 tw=78:
```

---

```bash
$ cd /path-to-your-vim/src
$ vim --clean --cmd "lang C" -S tabline_example.vim main.c
```

<img width="903" height="582" alt="Image" src="https://github.com/user-attachments/assets/3953cf39-2a59-4d5b-b515-9f4841ba63e0" />

---

```
:tabnew
:tabprev
```
<img width="903" height="582" alt="Image" src="https://github.com/user-attachments/assets/26a3295d-73cf-40f1-8c43-63fc7acd6898" />

**Comment by marcotrosi** [comment]
@h-east thanks for sharing this snippet, it's certainly a nice workaround to be considered in case we never get a global statusline. But then I lose the normal tabline, it's not at the bottom, I think it's not multi-line capable, what happens if someone uses a GUI-Vim, and what if someone uses tabpanel?

**Comment by h-east** [comment]
>it's certainly a nice workaround to be considered in case we never get a global statusline. 

The example I provided isn't a "workaround." It's something that can actually be done.
That's why I've only configured and displayed the tabline.

> But then I lose the normal tabline, it's not at the bottom,
> I think it's not multi-line capable, what happens if someone uses a GUI-Vim, and what if someone uses tabpanel?

As explained above.
And Why don't you try it yourself?
